### PR TITLE
Normalize @Path values by always stripping leading and trailing /

### LINKED ISSUE #479: support @Path w/o leading "/'
Most other REST frameworks allow both `@Path("hello")` and `@Path("/hello")` but MSF4J only recognizes the latter.  Also, JAX-RS spec only talks about relative path parameters and it's example always use the non-/ variant.

### PR DESCRIPTION
**Pull request by nkiesel**
JAX-RS spec says that leading or trailing spaces of path values should be
ignored.  This allows to use either `@Path("hello")` or `@Path("/hello")`.
Implement by trimming slashes and also replacing multiple / with a single one.

## Purpose
Resolves #479 
## Approach
Add `Utils.normalizePath` and call where necessary

## Automation tests
 - Unit tests 
   > Modified and tested with "helloworld" sample

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: core/src/main/java/org/wso2/msf4j/internal/router/MicroserviceMetadata.java [hunk 1]
```diff
@@ -181,7 +178,7 @@ private boolean isHttpMethodAvailable(Method method) {
                 throw new HandlerException(Response.Status.METHOD_NOT_ALLOWED, uri);
             } else {
                 throw new HandlerException(Response.Status.NOT_FOUND,
-                        String.format("Problem accessing: %s. Reason: Not Found", uri));
+                        String.format("Problem accessing: %s. Reason2: Not Found", uri));
```
Comment (vidurananayakkara) [inline core/src/main/java/org/wso2/msf4j/internal/router/MicroserviceMetadata.java]: What would be the reason for changing "Reason" to "Reason2"?

Comment (nkiesel) [reply core/src/main/java/org/wso2/msf4j/internal/router/MicroserviceMetadata.java]: Oops, leftover from testing when I was trying to find out which of the two identical error messages was actually producing the client side message.  I will remove that.

File: core/src/main/java/org/wso2/msf4j/util/Utils.java [hunk 2]
```diff
@@ -114,4 +114,38 @@ public static int getIntersection(Set set1, Set set2) {
...
+                continue;
+            }
+
+            if (p != i) {
+                chars[p] = c;
+            }
+            ++p;
+        }
+
+        while (p > 0 && chars[p - 1] == '/') {
```
Comment (nkiesel) [inline core/src/main/java/org/wso2/msf4j/util/Utils.java]: This could actually be coded as `if (p > 0...` because the code further up already makes sure we have at most one '/' left at the end of the string

File: core/src/main/java/org/wso2/msf4j/util/Utils.java [hunk 3]
```diff
@@ -114,4 +114,38 @@ public static int getIntersection(Set set1, Set set2) {
...
         return cloneSet2.size();
     }
+
+    /**
+     * Get the path without leading or trailing / and with multiple / reduced to one /
+     *
+     * @param path The path
+     * @return normalized path
+     */
+    public static final String normalizePath(String path) {
```
Comment (vidurananayakkara) [inline core/src/main/java/org/wso2/msf4j/util/Utils.java]: The logic here is too complex IMO. What do you think if we can simplify this code as below?

```java
    public static String normalizePath(String path) {

        // Replace multiple '/' characters with a single '/'
        path = path.replaceAll("[//]+", "/");

        // Remove leading '/'
        if (path.startsWith("/")) {
            path = path.substring(1, path.length());
        }

        // Remove trailing '/'
        if (path.endsWith("/")) {
            path = path.substring(0, path.length() - 1);
        }

        return path;
    }
```

File: core/src/main/java/org/wso2/msf4j/util/Utils.java [hunk 4]
```diff
@@ -114,4 +114,38 @@ public static int getIntersection(Set set1, Set set2) {
         cloneSet2.retainAll(cloneSet1);
         return cloneSet2.size();
     }
+
+    /**
+     * Get the path without leading or trailing / and with multiple / reduced to one /
+     *
+     * @param path The path
```
Comment (vidurananayakkara) [inline core/src/main/java/org/wso2/msf4j/util/Utils.java]: Would it be clear if we say, `resource path`?

File: core/src/main/java/org/wso2/msf4j/util/Utils.java [hunk 5]
```diff
@@ -114,4 +114,38 @@ public static int getIntersection(Set set1, Set set2) {
         cloneSet2.retainAll(cloneSet1);
         return cloneSet2.size();
     }
+
+    /**
+     * Get the path without leading or trailing / and with multiple / reduced to one /
```
Comment (vidurananayakkara) [inline core/src/main/java/org/wso2/msf4j/util/Utils.java]: Shall we add single quotes for the slashes (ex: '/') since it will be more readable? WDYT?

**Comment by nkiesel** [comment]
I will simplify the code and add some test cases. Stay tuned.

File: core/src/main/java/org/wso2/msf4j/util/Utils.java (hunk 5 above)
Comment (nkiesel) [reply core/src/main/java/org/wso2/msf4j/util/Utils.java]: Will do that.

File: core/src/main/java/org/wso2/msf4j/util/Utils.java (hunk 3 above)
Comment (nkiesel) [reply core/src/main/java/org/wso2/msf4j/util/Utils.java]: Your `"[//]+"` matches the same as `"/+"`, did you mean `"//+"`? All comes out to the same behavior though.

We could simplify that even further to
```
return path.replaceAll("/+", "/").replaceAll("^/", "").replaceAll("/$", "");
```
or (a bit more cryptic, but one fewer call)
```
return path.replaceAll("/+", "/").replaceAll("^/|/$", "");
```
Slightly less efficient than the original code, but much more readable.  I would be happy to change it to that.

File: core/src/main/java/org/wso2/msf4j/util/Utils.java (hunk 4 above)
Comment (nkiesel) [reply core/src/main/java/org/wso2/msf4j/util/Utils.java]: You want to just change the comment or also the method name?

**Comment by nkiesel** [comment]
Updated code based on the review comments.  You want me to squash all of this into a single commit and open a new merge request for that?

File: core/src/test/java/org/wso2/msf4j/util/UtilsTest.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+import static org.testng.AssertJUnit.assertEquals;
+
+/**
+ * Tests for utility class.
+ */
+public class UtilsTest {
+
+    @Test
+    public void testNormalizePath() {
+        assertEquals("hello", Utils.normalizePath("hello"));
```
Comment (vidurananayakkara) [inline core/src/test/java/org/wso2/msf4j/util/UtilsTest.java]: Instead of using repetitive assert statements, you can use a data provider.

**Example:**

```java
    @DataProvider(name = "paths")
    public Object[][] paths() {
        return new Object[][]{
                {"/path", "path"},
                {"path/", "path"},
                {"/path/", "path"},
                {"//path", "path"},
                {"///path", "path"},
                {"///path///", "path"},
                {"path1/path2", "path1/path2"},
                {"path1//path2", "path1/path2"},
                {"path1///path2", "path1/path2"},
                {"//path1//path2//", "path1/path2"},
                {"path1/path2/path3/path4", "path1/path2/path3/path4"},
                {"path1//path2//path3//path4", "path1/path2/path3/path4"},
                {"//path1//path2//path3//path4//", "path1/path2/path3/path4"},
                {"", ""},
                {"/", ""},
                {"//", ""},
                {"///", ""}
        };
    }

    /**
     * Test case to test normalization of paths.
     *
     * @param path         test path
     * @param expectedPath expected path after normalising the test path
     */
    @Test(dataProvider = "paths")
    public void normalizePathTest(String path, String expectedPath) {
        String normalizedPath = Utils.normalizePath(path);
        assertEquals(normalizedPath, expectedPath);
    }
```

This will run the above test case multiple times for the given dataset.

WDYT?

File: core/src/test/java/org/wso2/msf4j/util/UtilsTest.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        assertEquals("h/e/ll/o", Utils.normalizePath("h//e//ll//o"));
+        assertEquals("h/e/ll/o", Utils.normalizePath("//h//e//ll//o//"));
+
+        assertEquals("", Utils.normalizePath(""));
+        assertEquals("", Utils.normalizePath("/"));
+        assertEquals("", Utils.normalizePath("//"));
+        assertEquals("", Utils.normalizePath("///"));
+        assertEquals("h", Utils.normalizePath("h"));
+        assertEquals("h", Utils.normalizePath("/h"));
+        assertEquals("h", Utils.normalizePath("h/"));
```
Comment (vidurananayakkara) [inline core/src/test/java/org/wso2/msf4j/util/UtilsTest.java]: I think we have already tested below 2 assertions in line 30 and 31. Hence the below 2 assertions are not required
```java
    assertEquals("h", Utils.normalizePath("/h"));
    assertEquals("h", Utils.normalizePath("h/"));
```

File: core/src/main/java/org/wso2/msf4j/internal/router/MicroserviceMetadata.java [hunk 8]
```diff
@@ -73,10 +73,7 @@ public MicroserviceMetadata(Map<String, Object> services) {
                     patternRouter.add(absolutePath, new HttpResourceModel(absolutePath, method, service, false));
                 } else if (Modifier.isPublic(method.getModifiers()) && method.getAnnotation(Path.class) != null) {
                     // Sub resource locator method
-                    String relativePath = method.getAnnotation(Path.class).value();
-                    if (relativePath.startsWith("/")) {
-                        relativePath = relativePath.substring(1);
-                    }
+                    String relativePath = Utils.normalizePath(method.getAnnotation(Path.class).value());
```
Comment (vidurananayakkara) [inline core/src/main/java/org/wso2/msf4j/internal/router/MicroserviceMetadata.java]: Shall we normalize in the patternRouter.add method only.

We already do a normalisation in the patternRouter.add method (1st 2 lines of the method are as shown below)

```java
        // replace multiple slashes with a single slash.
        String path = source.replaceAll("/+", PATH_SLASH);

        path = (path.endsWith(PATH_SLASH) && path.length() > 1)
                ? path.substring(0, path.length() - 1) : path;
```
Instead of this we can use the new normalize method.

Then we do not need to normalise here and it will gain us a little bit of performance :)

File: core/src/main/java/org/wso2/msf4j/internal/router/MicroserviceMetadata.java [hunk 9]
```diff
@@ -105,7 +102,7 @@ public void addMicroserviceMetadata(final Object service, String basePath) {
             if (Modifier.isPublic(method.getModifiers()) && isHttpMethodAvailable(method)) {
                 String relativePath = "";
                 if (method.getAnnotation(Path.class) != null) {
-                    relativePath = method.getAnnotation(Path.class).value();
+                    relativePath = Utils.normalizePath(method.getAnnotation(Path.class).value());
```
Comment (vidurananayakkara) [inline core/src/main/java/org/wso2/msf4j/internal/router/MicroserviceMetadata.java]: Refer to the above comment

File: core/src/main/java/org/wso2/msf4j/internal/router/MicroserviceMetadata.java (hunk 8 above)
Comment (nkiesel) [reply core/src/main/java/org/wso2/msf4j/internal/router/MicroserviceMetadata.java]: Hmm, then why did the original code tried to strip a leading "/'?  Also looks to me the relativePath comes directly from the annotation (e.g. `@Path("/sub")`), so I don't understand your "is already normalized".  Seems I have to read the code further...

**Comment by vidurananayakkara** [comment]
Hi @nkiesel ,

IMO we should normalise the request URI in the beginig of class `org.wso2.msf4j.internal.router.MicroserviceMetadata`, method `getMatchedDestination` (line 201) as well.

File: core/src/main/java/org/wso2/msf4j/internal/router/MicroserviceMetadata.java [hunk 10]
```diff
@@ -145,7 +142,7 @@ private boolean isHttpMethodAvailable(Method method) {
                                                                          List<String> acceptHeader)
             throws HandlerException {
         try {
-            String path = URI.create(uri).normalize().getPath();
+            String path = Utils.normalizePath(URI.create(uri).normalize().getPath());
```
Comment (vidurananayakkara) [inline core/src/main/java/org/wso2/msf4j/internal/router/MicroserviceMetadata.java]: We do not need to normalise the path here since we already do normalise in `patternRouter.getDestinations` method. We should ideally also normalise the path in `getMatchedDestination` method in `MicroserviceMetaData` class (added a comment separately regarding this). WDYT?

File: core/src/test/java/org/wso2/msf4j/util/UtilsTest.java (hunk 7 above)
Comment (nkiesel) [reply core/src/test/java/org/wso2/msf4j/util/UtilsTest.java]: I like to treat these tests as "black box" tests and thus trying to test edge conditions here like empty path, path only consisting one or more slashes, or path of length 1.  Of course with the current implementation, there is no chance that these will fail. But my original optimized code did array index ops and for that off-by-one errors could have been possible.

I can add some more comments to the test cases explaining if that helps.

**Comment by vidurananayakkara** [comment]
Hi @nkiesel ,

Added few new comments in the code as well :)

File: core/src/test/java/org/wso2/msf4j/util/UtilsTest.java (hunk 6 above)
Comment (nkiesel) [reply core/src/test/java/org/wso2/msf4j/util/UtilsTest.java]: I chose "hello" because of the "helloWorld" example, but I'm also ok with your approach.  will change and push.

File: core/src/main/java/org/wso2/msf4j/internal/router/MicroserviceMetadata.java (hunk 8 above)
Comment (vidurananayakkara) [reply core/src/main/java/org/wso2/msf4j/internal/router/MicroserviceMetadata.java]: I think you misunderstood my comment. patternRouter.add method will normalise the path before proceeding with its logic (with this PR using the new normalising method). The original code which was here for stripping the leading '/' actually is not necessary previously as well and should have been removed earlier.
Since we are anyway normalising the path in patternRouter.add method, we do not need to normalise the path in line 76 at all.

WDYT?

File: core/src/test/java/org/wso2/msf4j/util/UtilsTest.java (hunk 7 above)
Comment (vidurananayakkara) [reply core/src/test/java/org/wso2/msf4j/util/UtilsTest.java]: Yes. You are right. A comment similar to "Testing the edge case of the path of length 1" would help for the above 2 assertions :)

File: core/src/main/java/org/wso2/msf4j/internal/router/PatternPathRouter.java [hunk 11]
```diff
@@ -66,12 +66,7 @@ public PatternPathRouter() {
      */
     public void add(final String source, final T destination) {
 
-        // replace multiple slashes with a single slash.
-        String path = source.replaceAll("/+", PATH_SLASH);
-
-        path = (path.endsWith(PATH_SLASH) && path.length() > 1)
-                ? path.substring(0, path.length() - 1) : path;
-
+        String path = Utils.normalizePath(source);
```
Comment (vidurananayakkara) [inline core/src/main/java/org/wso2/msf4j/internal/router/PatternPathRouter.java]: Remove blank line (line 70)

File: core/src/main/java/org/wso2/msf4j/internal/router/PatternPathRouter.java [hunk 12]
```diff
@@ -159,8 +154,7 @@ public void setName(String name) {
      */
     public List<RoutableDestination<T>> getDestinations(String path) {
 
-        String cleanPath = (path.endsWith(PATH_SLASH) && path.length() > 0)
-                ? path.substring(0, path.length() - 1) : path;
+        String cleanPath = Utils.normalizePath(path);
```
Comment (vidurananayakkara) [inline core/src/main/java/org/wso2/msf4j/internal/router/PatternPathRouter.java]: Remove blank line (line 158)

File: core/src/main/java/org/wso2/msf4j/internal/router/MicroserviceMetadata.java (hunk 10 above)
Comment (nkiesel) [reply core/src/main/java/org/wso2/msf4j/internal/router/MicroserviceMetadata.java]: Decided to convert code that did open-coded "/" removal with calls to `normalizePath`.  That is done now.

**Comment by nkiesel** [comment]
Hi,
this seems to have stalled.  Is the ball in my court? If so: please tell me what else you want to see changed.

**Comment by vidurananayakkara** [comment]
Hi @nkiesel ,

The PR looks fine now :)

However,

- `HttpMethodInfo` and `MicroservicesRegistryImpl` classes are missing the import`org.wso2.msf4j.util.Utils`
- There are integration test failures which needs to be addressed.

Can you resolve these issues as well?

**Comment by nkiesel** [comment]
Missing imports: sure (I guess I did not push everything)
integration test failures: can you point me to a concrete case that fails and how to trigger (so that I don't start going down the wrong path).

Currently swamped at work, so might be early next week before I can work again on this.

**Comment by vidurananayakkara** [comment]
Hi @nkiesel ,

The following changes need to be addressed in this PR,

- `HttpMethodInfo` and `MicroservicesRegistryImpl` classes are missing the importorg.wso2.msf4j.util.Utils.
- Unit tests are failing after the changes in the PR (run 'mvn clean install`). These need to be corrected.

I will be closing this PR since these changes are not addressed as of yet. It is appreciated if you could open a new PR with these changes :)
